### ISSUE TITLE
problem with dimensionality for non-periodic boundary conditions in lammps.base

### ISSUE THREAD

**Issue by JDaubin** [aiidaplugins/aiida-lammps#121]
Hi everyone,
in `write_structure_block` function the `get_dimensionality` method from AiiDA's `StructureData` class is used to compute dimensionality and write Lammps input files. The problem is that `get_dimensionality` actually computes the number of periodic boundary conditions which is not Lammps' dimensionality if I understand correctly.

For example the following structure leads to a Lammps error :
```
cell = [[1,0,0],
        [0,1,0],
        [0,0,1]]
structure = orm.StructureData(cell = cell, pbc = [True,True,False])
```
because `write_structure_block` writes in the Lammps input file : 
```
dimension 2
boundary p p f
```
instead of : 
```
dimension 3
boundary p p f
```

**Comment by JDaubin** [comment]
@JPchico thanks for your reply! I think the problem is the other way around in this case: we would like a 3D simulation with a non-periodic boundary condition in LAMMPS but AiiDA assumes only 2 dimensions.

I suppose the second solution will be necessary at some point to allow for more boundary condition types than periodic and fixed but, for simpler cases wouldn't it be possible to compute the dimension for LAMMPS as the length of the cell list from the `StructureData`?

```
dim = len(structure.cell)
```

**Comment by JPchico** [comment]
Hi @JDaubin
Okay I understand now, I think that is not possible, as the cell in AiiDA always has three lattice vectors. So just modifying the size of the cell would not work. In some programs one can set that one of the vectors is equal to zero, this is the interpreted as basically the size is given by the cartesian coordinates along that direction.

I think that the issue is that in the aiida sense a 3D structure by definition has periodic boundary conditions in each direction. This I think is the result from the StructureData being built for mostly DFT calculations. I think this is actually something that maybe one could raise in the [aiida discourse ](https://aiida.discourse.group/).

I think that for the current implementation of the StructureData one needs to add the overwrites.
Maybe there is another way without it, but I cannot think about it right now.

**Comment by JPchico** [comment]
I made a discourse [post](https://aiida.discourse.group/t/dimensionality-of-structures-in-aiida/368) raising this issue. Feel free to write anything that I missed @JDaubin . I also have made a [branch](https://github.com/JPchico/aiida-lammps/tree/121-problem-with-dimensionality-for-non-periodic-boundary-conditions-in-lammpsbase) in my fork dealing with this, I'll be testing it soon and see if it works to solve this issue.

**Comment by mbercx** [comment]
Thanks for raising @JDaubin and pointing us to this issue @JPchico! We're currently in the process of redesigning the `StructureData`, so understanding more use cases is very valuable. I can see how the dimensionality of a system is not necessarily directly determined by the periodic boundary conditions. One system I can imagine is a 3D metal in which you are looking at heat transport in one direction.

Just to understand, what problem are you studying here? What are the boundary conditions in the 3rd direction? How many different types of boundary conditions does LAMMPS support?

**Comment by JDaubin** [comment]
hi @mbercx and @JPchico, thanks for the work done and your interest in this issue!

Unfortunately I am no expert neither in using LAMMPS nor molecular dynamics so I won't be able to give a detailed explanation of the use case. I am in the process of automating with AiiDA an existing workflow which uses LAMMPS to study dislocations in metals. In this stage LAMMPS is used for the relaxation of atoms between two plates which are "glued together" so the boundary atoms along the z direction are fixed with a zero velocity. For more context and a better explanation, the full workflow and this specific step are described in the 2.1.3 section of  [L M Dupuy et al 2024 Modelling Simul. Mater. Sci. Eng. 32 035015](https://iopscience.iop.org/article/10.1088/1361-651X/ad278a/pdf).

LAMMPS supports 4 kinds of [boundary conditions](https://docs.lammps.org/boundary.html):
- periodic
- non-periodic and fixed
- non-periodic and shrink-wrapped
- non-periodic and shrink-wrapped with a minimum value

**Comment by mbercx** [comment]
Great, thanks for the explanation @JDaubin. We'll consider the use case in our updated version of the `StructureData`, but that'll still take some time. So for now it would be better to add one of the workarounds that @JPchico suggested.

**Comment by JPchico** [comment]
@JDaubin I have made a fork in which the overrides to the dimension and the boundary types can be passed in the parameters dictionary directly. Of this way it should be possible to handle the situation that you have. I'll add some notes in the documentation about this and then I'll make a PR to solve this issue.

**Comment by JDaubin** [comment]
Thanks a lot @JPchico, I just tested the new branch and it works as expected for my use case. I only have 2 comments:
- in the json schema I think `structure.dimension` could be restricted to 2,3 instead of 1,2,3 because [LAMMPS](https://docs.lammps.org/dimension.html) doesn't allow 1 for this command.
- I was surprised with the requirement of python 3.9 instead of 3.8.

**Comment by JPchico** [comment]
No worries, thanks for the comment on dimension 1 will change that.
Yeah this is because the latest release of aiida-core has dropped python 3.8 that is why. Is that a problem?

**Comment by JDaubin** [comment]
No, actually I have been using python 3.8 only because of #116 but now that you fixed this issue it isn't a problem anymore.

**Comment by JPchico** [comment]
Aha, yes that is fixed in the develop branch, will be merging that once the PRs are approved. The issue was not actually python 3.8 if not the jsonschema version that has changed how some validation happens and was causing the problems. I have about changing to pydantic to avoid this problem, but that is a major change that I think should be taken for a new major version.

**Comment by JPchico** [comment]
Fixed via #122
